fendermint/eth/api/examples/query_blockhash.rs

```diff
@@ -0,0 +1,168 @@
+
+#[allow(dead_code)]
+mod common;
+
+// Generate a statically typed interface for the contract.
+abigen!(QueryBlockhash, "../../testing/contracts/QueryBlockhash.abi");
+
+const QUERYBLOCKHASH_HEX: &'static str =
+    include_str!("../../../testing/contracts/QueryBlockhash.bin");
+
+#[derive(Parser, Debug)]
+pub struct Options {
+    /// The host of the Fendermint Ethereum API endpoint.
+    #[arg(long, default_value = "127.0.0.1", env = "FM_ETH__LISTEN__HOST")]
+    pub http_host: String,
+
+    /// The port of the Fendermint Ethereum API endpoint.
+    #[arg(long, default_value = "8545", env = "FM_ETH__LISTEN__PORT")]
+    pub http_port: u32,
+
+    /// Secret key used to deploy the contract.
+    ///
+    /// Assumed to exist with a non-zero balance.
+    #[arg(long, short)]
+    pub secret_key: PathBuf,
+
+    /// Path to write the contract metadata to.
+    #[arg(long, short)]
+    pub out: Option<PathBuf>,
+
+    /// Enable DEBUG logs.
+    #[arg(long, short)]
+    pub verbose: bool,
+}
+
+impl Options {
+    pub fn log_level(&self) -> Level {
+        if self.verbose {
+            Level::DEBUG
+        } else {
+            Level::INFO
+        }
+    }
+
+    pub fn http_endpoint(&self) -> String {
+        format!("http://{}:{}", self.http_host, self.http_port)
+    }
+}
+
+/// See the module docs for how to run.
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let opts: Options = Options::parse();
+
+    tracing_subscriber::fmt()
+        .with_max_level(opts.log_level())
+        .init();
+
+    let provider = Provider::<Http>::try_from(opts.http_endpoint())?;
+
+    run_http(provider, &opts).await?;
+
+    Ok(())
+}
+
+async fn run<C>(provider: &Provider<C>, opts: &Options) -> anyhow::Result<()>
+where
+    C: JsonRpcClient + Clone + 'static,
+{
+    let from = TestAccount::new(&opts.secret_key)?;
+
+    tracing::info!(from = ?from.eth_addr, "ethereum address");
+    tracing::info!("deploying QueryBlockhash");
+
+    let bytecode =
+        Bytes::from(hex::decode(QUERYBLOCKHASH_HEX).context("failed to decode contract hex")?);
+    let abi: Abi = QUERYBLOCKHASH_ABI.clone();
+
+    let chain_id = provider.get_chainid().await?;
+
+    let mw = make_middleware(provider.clone(), chain_id.as_u64(), &from)
+        .context("failed to create middleware")?;
+
+    let mw = Arc::new(mw);
+
+    let factory = ContractFactory::new(abi, bytecode.clone(), mw.clone());
+    let deployer = factory.deploy(())?;
+
+    let (contract, _deploy_receipt): (_, TransactionReceipt) =
+        deployer
+            .send_with_receipt()
+            .await
+            .context("failed to send deployment")?;
+
+    tracing::info!(addr = ?contract.address(), "QueryBlockhash deployed");
+
+    let contract = QueryBlockhash::new(contract.address(), contract.client());
+
+    // check that the blockhash returned by the contract matches the one returned by tendermint
+    for epoch in 1..=5 {
```

Comment:
Is there possibility that this could be flaky?

Reply:
How do you mean?

Reply:
I mean could it happen that when this test runs, those epochs have not happened?

Reply:
I doubt it because we have `cometbft-wait` that surely gives time for 5 blocks to be created, but as I commented further down the contract might have forgotten the first 5.